test_common/harness/kernelHelpers.cpp

```diff
@@ -809,6 +810,76 @@ int create_single_kernel_helper(cl_context context,
                                 const char *buildOptions,
                                 const bool openclCXX)
 {
+    // For the logic that automactially adds -cl-std it is much cleaner if the
+    // build options have RAII. This buffer will store the potentially updated
+    // build options, in which case buildOptions will point at the string owned
+    // by this buffer.
+    std::string build_options_internal{ buildOptions ? buildOptions : "" };
+
+    // Check the build options for the -cl-std option.
+    if (!buildOptions || !strstr(buildOptions, "-cl-std"))
+    {
+        // If the build option isn't present add it using the latetest OpenCL-C
+        // version supported by the device. This allows calling code to force a
+        // particular CL C version if it is required, but also means that
+        // callers need not specify a version if they want to assume the most
+        // recent CL C.
+
+        // First get all the devices in the context and find the minumum
+        // supported OpenCL C version.
+        size_t devices_size_in_bytes{};
+        auto error = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, nullptr,
+                                      &devices_size_in_bytes);
+        test_error(error, "clGetDeviceInfo failed for CL_CONTEXT_DEVICES");
+        std::vector<cl_device_id> devices(devices_size_in_bytes
+                                          / sizeof(cl_device_id));
+        error =
+            clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size_in_bytes,
+                             devices.data(), nullptr);
+        auto current_version = get_device_cl_version(devices[0]);
```

Comment:
I think this is fine for now, but at some point the device OpenCL C version could be different from the device OpenCL version.  When this occurs we should consider adding a helper function that gets the latest supported OpenCL C version, either by querying [CL_DEVICE_OPENCL_C_ALL_VERSIONS](https://bashbaug.github.io/OpenCL-Docs/html/OpenCL_API.html#CL_DEVICE_OPENCL_C_ALL_VERSIONS) (when supported), or falling back to [CL_DEVICE_OPENCL_C_VERSION](https://bashbaug.github.io/OpenCL-Docs/html/OpenCL_API.html#CL_DEVICE_OPENCL_C_VERSION) otherwise.

Reply:
Thanks for your feedback @bashbaug,  is really appreciated! Good point, I considered doing this but I figured that since `CL_DEVICE_OPENCL_C_ALL_VERSIONS` is missing before OpenCL-3.0 it wasn't an ideal way to go about querying the OpenCL C version. With regards to `CL_DEVICE_OPENCL_C_VERSION` as far as I understand, this returns the highest backwards compatible OpenCL C version, so for a 3.0 driver which optionally doesn't support some of the 2.X features, this would be forced to return 1.2, and I don't know if that is the behavior we want here since we are looking for the latest version supported by any device, not the latest backward compatible version?

Reply:
Perhaps maybe a good place to start would be to add a function getting the latest OpenCL C version using the same method as now (hard coded based on the device version), then we can more easily update this in the future using the correct device queries?